### PR TITLE
Add peakLocalMemory to query stats

### LINKED ISSUE #4562: Report max node memory reserved for each query
There are currently two memory limits that can be configured per query:
1) Max memory consumed globally
2) Max memory consumed on a single node

We currently have a stat to report the max memory consumed by a query globally, but not one to track the memory consumed on a single node (across all nodes). This latter one is an important metric to track since queries will fail if the limit is ever hit. This will allow users to see how close they are coming to hitting that threshold on a per query basis.

Since per-node memory thresholds are monitored exactly (vs. the eventually consistent model of the global memory tracking), we need to also report the per node memory consumed by a query inline with the main query stats.

### PR DESCRIPTION
**Pull request by rongrong**
Resolves #4562

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java [hunk 1]
```diff
@@ -343,6 +345,7 @@ public QueryInfo getQueryInfo(Optional<StageInfo> rootStage)
             cumulativeMemory += stageStats.getCumulativeMemory();
             totalMemoryReservation += stageStats.getTotalMemoryReservation().toBytes();
             peakMemoryReservation = getPeakMemoryInBytes();
+            peakLocalMemoryReservation = max(peakLocalMemoryReservation, stageStats.getPeakLocalMemoryReservation().toBytes());
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java]: Use Math.max

File: presto-main/src/main/java/com/facebook/presto/execution/StageStateMachine.java [hunk 2]
```diff
@@ -303,13 +306,13 @@ public StageInfo getStageInfo(Supplier<Iterable<TaskInfo>> taskInfosSupplier, Su
...
                 succinctBytes(totalMemoryReservation),
                 succinctBytes(peakMemoryReservation),
+                succinctBytes(peakLocalMemoryReservation),
                 succinctDuration(totalScheduledTime, NANOSECONDS),
                 succinctDuration(totalCpuTime, NANOSECONDS),
                 succinctDuration(totalUserTime, NANOSECONDS),
                 succinctDuration(totalBlockedTime, NANOSECONDS),
                 fullyBlocked && runningTasks > 0,
                 blockedReasons,
-
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/execution/StageStateMachine.java]: Revert

File: presto-spi/src/main/java/com/facebook/presto/spi/eventlistener/QueryStatistics.java [hunk 3]
```diff
@@ -53,10 +55,11 @@ public QueryStatistics(
         this.queuedTime = requireNonNull(queuedTime, "queuedTime is null");
         this.analysisTime = requireNonNull(analysisTime, "analysisTime is null");
         this.distributedPlanningTime = requireNonNull(distributedPlanningTime, "distributedPlanningTime is null");
-        this.peakMemoryBytes = requireNonNull(peakMemoryBytes, "peakMemoryBytes is null");
```
Comment (raghavsethi) [inline presto-spi/src/main/java/com/facebook/presto/spi/eventlistener/QueryStatistics.java]: You should fix these in a separate commit.

File: presto-main/src/main/java/com/facebook/presto/operator/TaskContext.java [hunk 4]
```diff
@@ -377,6 +377,7 @@ public TaskStats getTaskStats()
                 completedDrivers,
                 cumulativeMemory.get(),
                 succinctBytes(memoryReservation.get()),
+                succinctBytes(queryContext.getPeakReservedMemory()),
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/operator/TaskContext.java]: Don't understand this at all - why is the task constructing it's stats using the queryContext?

File: presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java [hunk 5]
```diff
@@ -58,6 +58,9 @@
     private long reserved;
 
     @GuardedBy("this")
+    private long peakReserved;
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java]: What is this? How is this relevant to the local memory change?

**Comment by rongrong** [comment]
Let's discuss in person tomorrow.
